Enhancement: Add models and methods for listing users, adding and removing dids from followed table

Ruturaj123 opened the pull request:
Enhancement: Add models and methods for listing users, adding and removing dids from followed table

Ruturaj123 committed:
Enhancement: Add models and methods for listing users, adding and removing dids from followed table

Enhancement: Travis fix

Enhancement: Syntax fix

Ruturaj123 committed:
Enhancement: Add api methods

Ruturaj123 committed:
Enhancement: Add new rest endpoint /follow for DIDs

Enhancement: Syntax fix

Ruturaj123 committed:
Enhancement: Create daemon for aggregating follow events

Enhancement: Syntax fix

mlassnig commented on lib/rucio/core/did.py:
```diff
@@ -1829,3 +1830,115 @@ def list_archive_content(scope, name, session=None):
+            new_did_followed = models.DidsFollowed(scope=did.scope, name=did.name, account=account,
+                                                   did_type=did.did_type, expired_at=did.expired_at)
+
+            new_did_followed.save(session=session, flush=True)
```
any reason why you flush to the DB every iteration? just curious

Ruturaj123 replied:
No Sir, not any specific reason as such. I will make the changes and flush after the loop is completed.

mlassnig commented on lib/rucio/core/did.py:
```diff
@@ -1829,3 +1830,115 @@ def list_archive_content(scope, name, session=None):
+    """
+    try:
+        query = session.query(models.DidsFollowed).filter_by(scope=scope, name=name).all()
+        users = list()
```
shouldn't be necessary to convert here? can you try without?

Ruturaj123 replied:
Without creating a list? Actually I created a list of dictionaries so that it can be rendered as json in the rest api

Ruturaj123 replied:
Here should I used yield instead? So that no need to create a list

mlassnig commented on lib/rucio/db/sqla/models.py:
```diff
@@ -1251,6 +1252,37 @@ class LifetimeExceptions(BASE, ModelBase):
+    account = Column(String(25))
+    did_type = Column(DIDType.db_type(name='FOLLOW_EVENTS_TYPE_CHK'))
+    event_type = Column(String(1024))
+    payload = Column(LargeBinary(255))
```
is the payload definition and size correct here?

Ruturaj123 replied:
Martin Sir had asked to keep the payload as a blob so that's why LargeBinary. And for the size I checked similar definitions in the models.py file and they had used 255 for LargeBinary. So I kept it the same.. Should it be increased?

bari12 replied:
For other tables we use `Column(Text)` for this. I would just do this here as well.

Ruturaj123 replied:
Okay Sir. I will change it to `Column(Text)`

Ruturaj123 committed:
Enhancement: Incorporate suggested changes

Enhancement: Make requested changes

bari12 commented on lib/rucio/core/did.py:
```diff
@@ -1829,3 +1830,115 @@ def list_archive_content(scope, name, session=None):
+        for did in dids:
+            did = session.query(models.DataIdentifier).filter_by(scope=did['scope'], name=did['name']).one()
+            new_did_followed = models.DidsFollowed(scope=did.scope, name=did.name, account=account,
+                                                   did_type=did.did_type, expired_at=did.expired_at)
```
expired_at is not really needed here.

bari12 commented on lib/rucio/api/did.py:
```diff
@@ -409,3 +410,57 @@ def list_archive_content(scope, name):
+    return did.get_users_following_did(name=name, scope=scope, session=session)
+
+
+def remove_did_from_followed(scope, name, session=None):
```
The core method has an account, this is needed here too.

bari12 commented on lib/rucio/core/did.py:
```diff
@@ -1829,3 +1830,115 @@ def list_archive_content(scope, name, session=None):
+        dids = session.query(models.DidsFollowed).filter_by(scope=scope, name=name).all()
+
+        for did in dids:
+            new_event = models.DidsFollowed(scope=scope, name=name, account=did.account,
```
I suppose this should be FollowEvents

bari12 commented on lib/rucio/daemons/follower/follower.py:
```diff
@@ -0,0 +1,116 @@
+
+    while not graceful_stop.is_set():
+        try:
+            accounts = session.query(FollowEvents.account).distinct()
```
Don't use SQLAlchemy in the daemon directly; Also do this via core calls.

bari12 commented on lib/rucio/db/sqla/models.py:
```diff
@@ -1251,6 +1252,37 @@ class LifetimeExceptions(BASE, ModelBase):
+    name = Column(String(NAME_LENGTH))
+    account = Column(String(25))
+    did_type = Column(DIDType.db_type(name='DIDS_FOLLOWED_TYPE_CHK'))
+    expires_at = Column(DateTime)
```
I don't think we need `expires_at` here.

bari12 commented on lib/rucio/db/sqla/models.py:
```diff
@@ -1251,6 +1252,37 @@ class LifetimeExceptions(BASE, ModelBase):
+                   CheckConstraint('SCOPE IS NOT NULL', name='DIDS_FOLLOWED_SCOPE_NN'),
+                   CheckConstraint('NAME IS NOT NULL', name='DIDS_FOLLOWED_NAME_NN'),
+                   CheckConstraint('DID_TYPE IS NOT NULL', name='DIDS_FOLLOWED_DID_TYPE_NN'),
+                   ForeignKeyConstraint(['account'], ['accounts.account'], name='DIDS_FOLLOWED_ACCOUNT_FK'))
```
Also add FK to scope name in dids

bari12 commented on lib/rucio/db/sqla/models.py:
```diff
@@ -1251,6 +1252,37 @@ class LifetimeExceptions(BASE, ModelBase):
+                   CheckConstraint('SCOPE IS NOT NULL', name='FOLLOW_EVENTS_SCOPE_NN'),
+                   CheckConstraint('NAME IS NOT NULL', name='FOLLOW_EVENTS_NAME_NN'),
+                   CheckConstraint('DID_TYPE IS NOT NULL', name='FOLLOW_EVENTS_DID_TYPE_NN'),
+                   ForeignKeyConstraint(['account'], ['accounts.account'], name='FOLLOW_EVENTS_ACCOUNT_FK'))
```
If you query this by account (like in the daemon) you will need an index on accoun too.

Ruturaj123 replied:
Index means a new `CheckConstraint` for account too right?

bari12 replied:
No, really an Index similar to other tables; Alternatively I suppose we could re-order the PK to account, scope, name as well.

Ruturaj123 replied:
Okay Sir.

bari12 commented on lib/rucio/web/rest/webpy/v1/did.py:
```diff
@@ -73,6 +75,7 @@
     '/new', 'NewDIDs',
     '/resurrect', 'Resurrect',
     '/list_dids_by_meta', 'ListByMeta',
+    '%s/follow/(.*)' % SCOPE_NAME_REGEXP, 'Follow',
```
only `%s/follow' 
There shouldn't be another parameter in the REST. the account should be part of the data

bari12 commented on lib/rucio/api/did.py:
```diff
@@ -409,3 +410,57 @@ def list_archive_content(scope, name):
+    :param name: The data identifier name.
+    :param session: The database session in use.
+    """
+    return did.remove_did_from_followed(scope=scope, name=name, session=session)
```
Please also add permission checks here

bari12 replied:
And on the above; Have a look how this is done with the issuer in other APIs.

bari12 reviewed (changes requested):
A few comments in the code; Please also have a look on the alembic migrate scripts; You need to provide an upgrade/downgrade script as well for the table generation.

bari12 commented:
Ah before I forget it, there is also things which came up in the syntax test in travis; Check the ones marked as Errors; (`E:...`)
It would also be good if you can add some test cases for adding/listing/removing followers etc.

Ruturaj123 commented:
Yes Sir. I will make all the changes soon

Ruturaj123 committed:
Enhancement: Add test cases for following mechanism

Ruturaj123 committed:
Enhancement: Add alembic migrate script

Enhancement: Travis fix

Enhancement: Travis fix

Enhancement: Correct test cases

Enhancement: Travis fix

Ruturaj123 committed:
Enhancement: Updated test cases

Enhancement: Syntax fix

cserf commented on lib/rucio/daemons/follower/follower.py:
```diff
@@ -0,0 +1,81 @@
+
+    while not graceful_stop.is_set():
+        # Create reports of events for all the accounts.
+        create_reports()
```
Each thread will send one report for all the accounts. Need partitioning

Ruturaj123 replied:
So how to go about doing the partitioning?

cserf commented on lib/rucio/core/did.py:
```diff
@@ -1829,3 +1830,157 @@ def list_archive_content(scope, name, session=None):
+                add_message('email', {'to': email,
+                                      'subject': 'Report of affected dataset(s)',
+                                      'body': body})
+
```
One should clean the events when the message was successfully sent, otherwise at the next iteration the same list of events will be sent. Another possibility is to keep all the events, but only send the recent ones, based on created_at.

cserf commented on lib/rucio/core/permission/atlas.py:
```diff
@@ -1042,3 +1045,35 @@ def perm_add_bad_pfns(issuer, kwargs):
+    """
+    return _is_root(issuer)\
+        or has_account_attribute(account=issuer, key='admin')\
+        or rucio.core.scope.is_scope_owner(scope=kwargs['scope'], account=issuer)\
```
I would remove the condition on the scope and add one on `kwargs['account] == issuer`

cserf commented on lib/rucio/core/permission/atlas.py:
```diff
@@ -1042,3 +1045,35 @@ def perm_add_bad_pfns(issuer, kwargs):
+    scopes = [did['scope'] for did in kwargs['dids']]
+    scopes = list(set(scopes))
+    for scope in scopes:
+        if not rucio.core.scope.is_scope_owner(scope, issuer):
```
Same comment as above (remove the condition on the scope and add one on `kwargs['account] == issuer`

cserf commented on lib/rucio/core/permission/cms.py:
```diff
@@ -875,3 +878,35 @@ def perm_add_bad_pfns(issuer, kwargs):
+    """
+    return _is_root(issuer)\
+        or has_account_attribute(account=issuer, key='admin')\
+        or rucio.core.scope.is_scope_owner(scope=kwargs['scope'], account=issuer)\
```
Same comment as above

cserf commented on lib/rucio/core/permission/cms.py:
```diff
@@ -875,3 +878,35 @@ def perm_add_bad_pfns(issuer, kwargs):
+    scopes = [did['scope'] for did in kwargs['dids']]
+    scopes = list(set(scopes))
+    for scope in scopes:
+        if not rucio.core.scope.is_scope_owner(scope, issuer):
```
Same comment as above

cserf commented on lib/rucio/core/permission/generic.py:
```diff
@@ -866,3 +869,35 @@ def perm_add_bad_pfns(issuer, kwargs):
+    """
+    return _is_root(issuer)\
+        or has_account_attribute(account=issuer, key='admin')\
+        or rucio.core.scope.is_scope_owner(scope=kwargs['scope'], account=issuer)\
```
Same comment as above

cserf commented on lib/rucio/core/permission/generic.py:
```diff
@@ -866,3 +869,35 @@ def perm_add_bad_pfns(issuer, kwargs):
+    scopes = [did['scope'] for did in kwargs['dids']]
+    scopes = list(set(scopes))
+    for scope in scopes:
+        if not rucio.core.scope.is_scope_owner(scope, issuer):
```
Same comment as above

cserf commented:
To keep in mind for you Martin : `delete_dids` should remove the rows in the followed table otherwise, we will have FK constraint violation

Ruturaj123 commented:
> To keep in mind for you Martin : `delete_dids` should remove the rows in the followed table otherwise, we will have FK constraint violation

So should I make changes in `delete_dids` for removing the rows from followed table?

bari12 commented:
Yes please.

Ruturaj123 committed:
Enhancement: Remove did from followed table in delete_dids

Ruturaj123 committed:
Enhancement: Modify permissions

Ruturaj123 committed:
Enhancement: Use heartbeat mechanism to create reports

Enhancement: Travis fix

Ruturaj123 committed:
Enhancement: Change Column type from String to InternalScopeString

Enhancement: Apply patch b881742

Enhancement: Syntax fix

Enhancement: Syntax fix

Ruturaj123 committed:
Enhancement: Update Alembic script

WebUI: Resolve Conflicts

bari12 commented on lib/rucio/api/did.py:
```diff
@@ -530,3 +531,67 @@ def list_archive_content(scope, name):
         yield api_update_return_dict(d)
+
+
+def add_did_to_followed(scope, name, account, session=None):
```
Please change all scopes and accounts to `InternalScope` and `InternalAccount`. E.g. have a look on the `create_did_sample` method in this module.

bari12 commented on lib/rucio/core/did.py:
```diff
@@ -1833,3 +1841,171 @@ def list_archive_content(scope, name, session=None):
+                                  'body': body})
+
+            # Clean up the events after creating the report
+            session.query(models.FollowEvents).filter_by(account=event.account).\
```
I don't think this deletion works, since you only processed a partition of the events of this account, but you are deleting all of them. @cserf suggested to delete the processed events when sending them.

bari12 commented on lib/rucio/db/sqla/migrate_repo/versions/277bec1ddba9_update_dids_followed_and_follow_events_.py:
```diff
@@ -0,0 +1,55 @@
+# Authors:
+# - Ruturaj Gujar, <ruturaj.gujar23@gmail.com>, 2019
+
+''' update dids_followed and follow_events table '''
```
Please merge this with the other migrate script. It's enough to do it in one 😄

bari12 reviewed (changes requested):
Hi Ruturaj, again a few comments. Sorry for the delay. Would be awesome if you could still integrate them! Thanks! 😄

Ruturaj123 committed:
Enhancement: Update migrate script and create_reports

Enhancement: Syntax fix

bari12 commented:
Looks good to me too. @cserf or @mlassnig  please also have a last look on this one.

cserf commented on lib/rucio/core/did.py:
```diff
@@ -1833,3 +1841,173 @@ def list_archive_content(scope, name, session=None):
+    query = session.query(models.FollowEvents)
+
+    # Use hearbeat mechanism to select a chunck of events based on the hashed account
+    if total_workers > 0:
```
Not blocking, but can switch to filter_thread_work

cserf commented on lib/rucio/core/did.py:
```diff
@@ -1833,3 +1841,173 @@ def list_archive_content(scope, name, session=None):
+            body += "Thank You."
+            # Get the email associated with the account.
+            email = session.query(models.Account.email).filter_by(account=account)
+            add_message('email', {'to': email,
```
Each thread will send a separate email. Can be improved.

cserf commented on lib/rucio/daemons/follower/follower.py:
```diff
@@ -0,0 +1,96 @@
+    else:
+        logging.info("starting follower threads")
+        # Run the follower daemon thrice a day
+        threads = [get_thread_with_periodic_running_function(28800, aggregate_events, graceful_stop) for i in range(threads)]
```
Interval should be parametrizable

cserf commented on lib/rucio/db/sqla/migrate_repo/versions/7541902bf173_add_didsfollowed_and_followevents_table.py:
```diff
@@ -0,0 +1,83 @@
+
+# Alembic revision identifiers
+revision = '7541902bf173'
+down_revision = '2cbee484dcf9'
```
Down revision must be changed (for you Martin)

cserf reviewed (approved):
A few non-blocking comments that can be addressed in a future PR. The only thing to change is the alembic revision number.
I would approve for now.
